modules/phase_field/src/utils/PolycrystalICTools.C

```diff
@@ -297,45 +300,78 @@ PolycrystalICTools::buildNodalGrainAdjacencyGraph(
-         *
-         * To do that we'll build a halo region based on these interface nodes. For now, we need to
-         * record
-         * the nodes inside of the grain and those outside of the grain.
+         * need to make sure that we build out a small buffer region to avoid literal "corner cases"
+         * where nodes on opposite corners of a QUAD end up with the same OP because those nodes are
+         * not nodal neighbors. To do that we'll build a halo region based on these interface nodes.
+         * For now, we need to record the nodes inside of the grain and those outside of the grain.
          */
-        adjacency_matrix[my_grain][their_grain] = 1;
+        adjacency_matrix(my_grain, their_grain) = 1;
     }
   }
 
   return adjacency_matrix;
 }
 
 std::vector<unsigned int>
-PolycrystalICTools::assignOpsToGrains(const AdjacencyGraph & adjacency_matrix,
+PolycrystalICTools::assignOpsToGrains(AdjacencyMatrix<Real> & adjacency_matrix,
                                       unsigned int n_grains,
-                                      unsigned int n_ops)
+                                      unsigned int n_ops,
+                                      const MooseEnum & coloring_algorithm)
 {
   Moose::perf_log.push("assignOpsToGrains()", "PolycrystalICTools");
 
   std::vector<unsigned int> grain_to_op(n_grains, GraphColoring::INVALID_COLOR);
 
-  if (!colorGraph(adjacency_matrix, grain_to_op, n_grains, n_ops, 0))
-    mooseError(
-        "Unable to find a valid Grain to op configuration, do you have enough op variables?");
+  // Use a simple backtracking coloring algorithm
+  if (coloring_algorithm == "bt")
+  {
+    if (!colorGraph(adjacency_matrix, grain_to_op, n_grains, n_ops, 0))
+      mooseError(
+          "Unable to find a valid Grain to op configuration, do you have enough op variables?");
+  }
+  else // PETSc Coloring algorithms
+  {
+#ifdef LIBMESH_HAVE_PETSC
+    const std::string & ca_str = coloring_algorithm;
+    Real * am_data = adjacency_matrix.rawDataPtr();
+    Moose::PetscSupport::colorAdjacencyMatrix(
+        am_data, n_grains, n_ops, grain_to_op, ca_str.c_str());
+#else
+    mooseError("Selected coloring algorithm requires PETSc");
+#endif
+  }
 
   Moose::perf_log.pop("assignOpsToGrains()", "PolycrystalICTools");
 
   return grain_to_op;
 }
 
+MooseEnum
+PolycrystalICTools::coloringAlgorithms()
+{
+  return MooseEnum("legacy bt jp power sl lf id greedy", "legacy");
+}
+
+std::string
+PolycrystalICTools::coloringAlgorithmDescriptions()
+{
+  return "The grain neighbor graph coloring algorithm to use. \"legacy\" is the original "
+         "algorithm "
+         "which does not guarantee a valid coloring. \"bt\" is a simple backtracking algorithm "
+         "which will produce a valid coloring but has potential exponential run time. The "
+         "remaining algorithms require PETSc but are recommended for larger problems (See "
```

Comment:
Yeah, I imagine so. In my mind, I'm still thinking about the case where Trilinos could be used even though we don't do this practically anymore.